serialization/case_objects.nim

```diff
@@ -0,0 +1,297 @@
+        `code`.add nnkExprColonExpr.newTree(ident `patchedNameStr`,
+          nnkWhenStmt.newTree(
+            nnkElifExpr.newTree(
+              newCall(ident "compiles", staticConvert),
+              nnkPar.newTree(staticConvert)),
+            nnkElse.newTree(`arg`[1])))
+      else:
+        `code`.add nnkExprColonExpr.newTree(`arg`[0], `arg`[1]))
+    fieldPairsCode.add nnkElifBranch.newTree(
+      quote do: `keyId` == `patchedNameStr`,
+      quote do:
+        when not isSupportedAsDiscriminator(`OrigTyp`):
+          const `keyParam` {.inject, used.} = `origNameStr`
+          template `valParam`: `OrigTyp` {.inject, used.} = `valId`
+        else:
+          {.error: `patchedNameStr` & " should not require patching".})
+  initCode.add nnkElse.newTree quote do:
+    `code`.add nnkExprColonExpr.newTree(`arg`[0], `arg`[1])
+  fieldPairsCode.add nnkElse.newTree quote do:
+    const `keyParam` {.inject, used.} = `keyId`
+    template `valParam`: untyped {.inject, used.} = `valId`
+
+  nnkTypeSection.newTree(
+    nnkTypeDef.newTree(
+      nskType.genSym "",
+      newEmptyNode(),
+      nnkStmtListType.newTree(
+        converterCode,
+        nnkTypeSection.newTree(typ),
+        accessorCode,
+        (quote do:
+          macro `initId`(
+              t: typedesc[`T`],
+              args: varargs[untyped]{nkArgList}): `T` {.used.} =
+            var `code` = nnkObjConstr.newTree(t)
+            for `arg` in args:
+              if `arg`.kind != nnkExprEqExpr:
+                error "arguments must be passed by key", `arg`
+              `initCode`
+            `code`
+
+          iterator `fieldPairsId`[t: `T`](
+              x: t,
+          ): tuple[key: string, val: RootObj] {.noSideEffect, used.} =
+            {.error: $t & " does not support `fieldPairs`; " &
+              "use `withFieldPairs` instead".}
+
+          template `doWithFieldPairs`(
+              x: `T`, `keyParam`: untyped, `valParam`: untyped,
+              body: untyped) {.used.} =
+            for `keyId`, `valId` in system.fieldPairs(x):
+              `fieldPairsCode`
+              body
+
+          iterator `fieldsId`[t: `T`](x: t): RootObj {.noSideEffect, used.} =
+            {.error: $t & " does not support `fields`; " &
+              "use `withFields` instead".}
+
+          template `doWithFields`(
+              x: `T`, `valParam`: untyped,
+              body: untyped) {.used.} =
+            `doWithFieldPairs`(x, _, `valParam`):
+              body
+
+          func `dollarId`(x: `T`): string {.used.} =
+            var
+              res = "("
+              didAdd = false
+            x.`doWithFieldPairs`(key, val):
+              if didAdd:
+                res &= ", "
+              res &= key & ": " & $val
+              didAdd = true
+            res &= ")"
+            res
+        ),
+        ident "void")))
+
+template withFieldPairs*(
+    x: auto, keyParam: untyped, valParam: untyped, body: untyped) =
```

Comment:
why is `x` specifically `auto`? similarly with `withFields`.

Reply:
and the reason why the template exists is that `fieldPairs` and `fields` cannot be overridden.
- they are treated specially by the compiler and transform the for loop in a way that's not accessible from external code
- have a fake signature that suggests that primitives are objects, so any override cannot provide same functionality